connection.go

```diff
@@ -858,6 +858,14 @@ func (c *Connection) checkExchanges() {
 func (c *Connection) close(fields ...LogField) error {
 	c.log.WithFields(fields...).Debug("Connection closing.")
 
+	// Remaining frames in sendCh must be drained and released,
+	// or we will leak frames
+	defer func() {
+		for len(c.sendCh) > 0 {
+			c.opts.FramePool.Release(<-c.sendCh)
+		}
+	}()
+
```

Comment:
This can act as a "best-effort" release frames, but there's many race conditions which can result in frames being written to `sendCh` after the drain here. In the past, TChannel used to `close(c.sendCh)` as part of closing a connection, but it was removed as there was many race conditions that could lead to the close happening while some other goroutine wrote to the channel, see
https://github.com/uber/tchannel-go/pull/388/commits/93ef5c112c8b321367ae52d2bd79396e2e874f31

The same race conditions that would have caused a panic when closing `sendCh` will likely cause this drain to "miss" frames from being released.

It's worth keeping in mind that a frame leak from the pool is *not* a memory leak -- when a connection is closed + removed, the GC will clean up the connection, `sencCh`, and all frames in that channel.

I think having the best-effort is OK, but the comment should indicate that it's not a complete solution.

If we want this logic, it should ideally happen as late in the lifecycle for a connection as possible, when we're as close to being confident that there's no other writers to `sendCh`. E.g., when the connection state changes to closed,
https://cs.github.com/uber/tchannel-go/blob/751f26a5d1d895affe75a1edd208d63fe3b7b109/connection.go?q=repo%3Auber%2Ftchannel-go+%22connectionStartClose%22#L847

(We could also keep the logic in `writeFrames`, and have `writeFrames` block waiting for `stopCh` before draining `sendCh`).

Reply:
good pointers! even if it's best-effort I think it's still useful since it'll help improve the signal of the tests since it won't mask leaks from other code paths.

moved the drain back into `writeFrames` and blocked on `<-stopCh`